File: sharedUI/src/commonMain/kotlin/com/dsankovsky/kmpclientplanner/ui/screens/settings/SettingsViewModel.kt
package com.dsankovsky.kmpclientplanner.ui.screens.settings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dsankovsky.kmpclientplanner.data.datastore.AppSettings
import com.dsankovsky.kmpclientplanner.domain.models.additional.ServiceType
import com.dsankovsky.kmpclientplanner.domain.usecases.ClearDatabaseUseCase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class SettingsViewModel(
    private val appSettings: AppSettings,
    private val getDatabaseUseCase: ClearDatabaseUseCase
) : ViewModel() {

    private val _state = MutableStateFlow(SettingsScreenState())
    val state = _state.asStateFlow()

    val event = MutableSharedFlow<SettingsScreenEvent>()

    fun handleActions(action: SettingsScreenAction) {
        when (action) {
            SettingsScreenAction.LoadData -> loadData()
            is SettingsScreenAction.OnServiceTypeSelected -> changeServiceType(action.serviceTypeIndex)
            SettingsScreenAction.DeleteAllData -> deleteAllData()
        }
    }

    private fun loadData() {
        viewModelScope.launch {
            val currentServiceType = appSettings.getServiceType()
            currentServiceType?.let { type ->
                _state.update {
                    it.copy(serviceType = type)
                }
            }
        }
    }

    private fun changeServiceType(serviceTypeIndex: Int) {
        val serviceType = ServiceType.entries[serviceTypeIndex]
        viewModelScope.launch {
            appSettings.saveServiceType(serviceType)
            _state.update {
                it.copy(serviceType = serviceType)
            }
        }
    }

    private fun deleteAllData() {
        viewModelScope.launch(Dispatchers.IO) {
            appSettings.clear()
            getDatabaseUseCase.clearDatabase()
            event.emit(SettingsScreenEvent.AllDataCleared)
        }
    }

}

File: iosApp/iosApp/Screens/Settings/SettingsViewModel.swift
import Foundation
import SharedLogic

/// iOS counterpart of `ui.screens.settings.SettingsViewModel`.
///
/// MVVM: the view observes `state`; intents come in through `handle(_:)`.
/// Dependencies (`AppSettings`, `ClearDatabaseUseCase`) are resolved from the
/// shared Koin graph via `IosDi`. Suspend functions are consumed as Swift
/// `async` thanks to Skie.
@MainActor
final class SettingsViewModel: ObservableObject {

    enum Action {
        case loadData
        case onServiceTypeSelected(ServiceType)
        case deleteAllData
    }

    @Published private(set) var state = SettingsScreenState()

    /// Emitted once when all data has been cleared (consumed by the view).
    @Published var event: SettingsScreenEvent?

    private let appSettings: AppSettings = IosDi.shared.appSettings
    private let clearDatabaseUseCase: ClearDatabaseUseCase = IosDi.shared.clearDatabaseUseCase

    func handle(_ action: Action) {
        switch action {
        case .loadData:
            loadData()
        case .onServiceTypeSelected(let serviceType):
            changeServiceType(serviceType)
        case .deleteAllData:
            deleteAllData()
        }
    }

    private func loadData() {
        Task {
            if let currentServiceType = try? await appSettings.getServiceType() {
                state.serviceType = currentServiceType
            }
        }
    }

    private func changeServiceType(_ serviceType: ServiceType) {
        Task {
            try? await appSettings.saveServiceType(serviceType: serviceType)
            state.serviceType = serviceType
        }
    }

    private func deleteAllData() {
        Task {
            try? await appSettings.clear()
            try? await clearDatabaseUseCase.clearDatabase()
            event = .allDataCleared
        }
    }
}
